partner_vat_unique/models/res_partner.py

```diff
@@ -0,0 +1,30 @@
+# -*- coding: utf-8 -*-
+# Copyright 2017 Grant Thornton Spain - Ismael Calvo <ismael.calvo@es.gt.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import api, models, _
+from odoo.exceptions import ValidationError
+from odoo.tools import config
+
+
+class ResPartner(models.Model):
+    _inherit = 'res.partner'
+
+    @api.constrains('vat')
+    def _check_vat_unique(self):
+        for record in self:
+            if record.parent_id:
+                continue
+            test_condition = (config['test_enable'] and
+                              not self.env.context.get('test_vat'))
+            if test_condition:
+                continue
+            results = self.env['res.partner'].search([
```

Comment:
Use search_count for performance.

https://www.odoo.com/documentation/10.0/reference/orm.html#odoo.models.Model.search_count